File: src/foam/dao/index/AltIndex.js
/**
 * @license
 * Copyright 2016 Google Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
  Provides for hetergenious indexes, where not all potential delegates
  of this AltIndex actually get populated for each instance. Each instance
  always populates an ID index, so it can serve queries even if no
  delegate indexes are explicitly added.

  Index: Alt[ID, TreeA, TreeB]
  IndexNodes: [id, a,b], [id, a], [id, b], [id, a], [id]
*/
foam.CLASS({
  package: 'foam.dao.index',
  name: 'AltIndex',
  extends: 'foam.dao.index.Index',

  requires: [
    'foam.dao.index.NoPlan',
    'foam.mlang.sink.NullSink',
  ],

  constants: {
    /** Maximum cost for a plan which is good enough to not bother looking at the rest. */
    GOOD_ENOUGH_PLAN: 10 // put to 10 or more when not testing
  },

  properties: [
    {
      /** delegate factories */
      name: 'delegates',
      factory: function() { return []; }
    },
    {
      /** factory quick lookup */
      name: 'delegateMap_',
      factory: function() { return {}; }
    },
  ],

  methods: [

    /** Returns smallest estimate from the delegates */
    function estimate(size, sink, skip, limit, order, predicate) {
      var cost = Number.MAX_VALUE;
      for ( var i = 0; i < this.delegates.length; i++ ) {
        cost = Math.min(
          cost,
          this.delegates[i].estimate(
            size, sink, skip, limit, order, predicate)
        );
      }
      return cost;
    },

    function toPrettyString(indent) {
      var ret = "";
      for ( var i = 0; i < this.delegates.length; i++ ) {
          ret += this.delegates[i].toPrettyString(indent + 1);
      }
      return ret;
    },

    function toString() {
      return 'Alt([' + (this.delegates.join(',')) + '])';
    },
  ]
});

foam.CLASS({
  package: 'foam.dao.index',
  name: 'AltIndexNode',
  extends: 'foam.dao.index.IndexNode',

  properties: [
    {
      /** the delegate instances for each Alt instance */
      class: 'Simple',
      name: 'delegates'
    },
  ],

  methods: [
    function init() {
      this.delegates = this.delegates || [ this.creator.delegates[0].createNode() ];
    },

    function addIndex(index) {
      // check for existing factory
      var dfmap = this.creator.delegateMap_;
      var indexKey = index.toString();
      if ( ! dfmap[indexKey] ) {
        this.creator.delegates.push(index);
        dfmap[indexKey] = index;
      } else {
        // ensure all tails are using the same factory instance
        index = dfmap[indexKey];
      }

      var newSubInst = index.createNode();
      this.delegates[0].plan(newSubInst).execute([], newSubInst);
      this.delegates.push(newSubInst);
    },

    function bulkLoad(a) {
      for ( var i = 0 ; i < this.delegates.length ; i++ ) {
        this.delegates[i].bulkLoad(a);
      }
    },

    function get(key) {
      return this.delegates[0].get(key);
    },

    function pickDelegate(order, cache) {
      // NOTE: this assumes one of the delegates is capable of ordering
      //  properly for a scan. We should not be asked for a select unless
      //  a previous estimate indicated one of our options was sorted properly.
      // NOTE: unbuilt portions of the index will be built immediately
      //  if picked for ordering.
      var delegates = this.delegates;
      if ( ! order ) return delegates[0];

      var c = cache[this];
      // if no cached index estimates, generate estimates
      // for each factory for this ordering
      if ( ! c ) {
        var nullSink = this.creator.NullSink.create();
        var dfs = this.creator.delegates;
        var bestEst = Number.MAX_VALUE;
        // Pick the best factory for the ordering, cache it
        for ( var i = 0; i < dfs.length; i++ ) {
          var est = dfs[i].estimate(1000, nullSink, undefined, undefined, order);
          if ( est < bestEst ) {
            c = dfs[i];
            bestEst = est;
          }
        }
        cache[this] = c;
      }

      // check if we have a delegate instance for the best factory
      for ( var i = 0; i < delegates.length; i++ ) {
        // if we do, it's the best one
        if ( delegates[i].creator === c ) return delegates[i];
      }

      // we didn't have the right delegate generated, so add and populate it
      // as per addIndex, but we skip checking the factory as we know it's stored
      var newSubInst = c.createNode();
      this.delegates[0].plan(newSubInst).execute([], newSubInst);
      this.delegates.push(newSubInst);

      return newSubInst;
    },


    function select(sink, skip, limit, order, predicate, cache) {
      // find and cache the correct subindex to use
      this.pickDelegate(order, cache)
        .select(sink, skip, limit, order, predicate, cache);
    },

    function put(newValue) {
      for ( var i = 0 ; i < this.delegates.length ; i++ ) {
        this.delegates[i].put(newValue);
      }
    },

    function remove(obj) {
      for ( var i = 0 ; i < this.delegates.length ; i++ ) {
        this.delegates[i].remove(obj);
      }
    },

    function plan(sink, skip, limit, order, predicate, root) {
      var bestPlan;
      for ( var i = 0 ; i < this.delegates.length ; i++ ) {
        var p = this.delegates[i].plan(sink, skip, limit, order, predicate, root);
        if ( p.cost <= this.creator.GOOD_ENOUGH_PLAN ) {
          bestPlan = p;
          break;
        }
        if ( ! bestPlan || p.cost < bestPlan.cost ) {
          bestPlan = p;
        }
      }
      if ( ! bestPlan ) {
        return this.creator.NoPlan.create();
      }
      return bestPlan;
    },

    function size() { return this.delegates[0].size(); },

    function toString() {
      return 'Alt([' + (this.creator.delegates.join(',')) + this.size() + '])';
    },
  ]
});

File: src/foam/dao/index/AltIndex.java
/**
 * @license
 * Copyright 2017 The FOAM Authors. All Rights Reserved.
 *     http://www.apache.org/licenses/LICENSE-2.0
 */
package foam.dao.index;

import foam.core.FObject;
import foam.dao.AbstractSink;
import foam.dao.Sink;
import foam.mlang.order.Comparator;
import foam.mlang.predicate.Predicate;
import java.util.ArrayList;

/** Note this class is not thread safe because ArrayList isn't thread-safe. Needs to be made safe by containment. **/
public class AltIndex
  extends AbstractIndex
{

  public final static int GOOD_ENOUGH_PLAN_COST = 10;

  protected ArrayList<Index> delegates_ = new ArrayList();

  public AltIndex(Index... indices) {
    for ( int i = 0 ; i < indices.length ; i++ )
      addIndex(null, indices[i]);
  }

  public Object addIndex(Object state, Index i) {
    delegates_.add(i);

    // No data to copy when just adding first index
    if ( delegates_.size() == 1 ) return state;

    // No state means no data to copy
    if ( state == null ) return state;

    // Copy all data from first index into new index, updating state
    final Object[] sa = toObjectArray(state);
    Sink sink = new AbstractSink() {
      public void put(Object obj, foam.core.Detachable sub) {
        sa[sa.length-1] = i.put(sa[sa.length-1], (FObject) obj);
      }
    };

    delegates_.get(0).planSelect(sa[0], sink, 0, Long.MAX_VALUE, null, null).select(sa[0], sink, 0, Long.MAX_VALUE, null, null);

    return sa;
  }

  protected Object[] toObjectArray(Object state) {
    Object[] s2 = new Object[delegates_.size()];

    if ( state != null ) {
      Object[] s1 = (Object[]) state;

      for ( int i = 0 ; i < s1.length ; i++ ) {
        s2[i] = s1[i];
      }
    }

    return s2;
  }

  public Object put(Object state, FObject value) {
    Object[] s = toObjectArray(state);

    for ( int i = 0 ; i < delegates_.size() ; i++ )
      s[i] = delegates_.get(i).put(s[i], value);

    return s;
  }


  public Object remove(Object state, FObject value) {
    Object[] s = toObjectArray(state);

    for ( int i = 0 ; i < delegates_.size() ; i++ )
      s[i] = delegates_.get(i).remove(s[i], value);

    return s;
  }

  public Object removeAll() {
    Object[] s = toObjectArray(null);

    for ( int i = 0 ; i < delegates_.size() ; i++ )
      s[i] = delegates_.get(i).removeAll();

    return s;
  }


  public FindPlan planFind(Object state, Object key) {
    if ( state == null ) return NotFoundPlan.instance();

    Object[] s         = (Object[]) state;
    FindPlan bestPlan  = NoPlan.instance();
    Object   bestState = null;

    for ( int i = 0 ; i < delegates_.size() ; i++ ) {
      FindPlan plan = delegates_.get(i).planFind(s[i], key);

      // only return the smallest cost plan
      if ( plan.cost() < bestPlan.cost() ) {
        bestPlan  = plan;
        bestState = s[i];
        if ( bestPlan.cost() <= GOOD_ENOUGH_PLAN_COST ) break;
      }
    }

    return new AltFindPlan(bestState, bestPlan);
  }

  public SelectPlan planSelect(Object state, Sink sink, long skip, long limit, Comparator order, Predicate predicate) {
    if ( state == null ) return NotFoundPlan.instance();

    Object[]   s                 = (Object[]) state;
    SelectPlan bestPlan          = NoPlan.instance();
    Object     bestState         = null;
    Predicate  originalPredicate = null;

    for ( int i = 0 ; i < delegates_.size() ; i++ ) {
      // ???: Why is this?
      // To keep the original predicate, because in our next operate the predicate will be changed
      if ( predicate != null ) {
        originalPredicate = (Predicate) ((FObject) predicate).deepClone();
      }

      SelectPlan plan = delegates_.get(i).planSelect(s[i], sink, skip, limit, order, originalPredicate);

      if ( plan.cost() < bestPlan.cost() ) {
        bestPlan  = plan;
        bestState = s[i];
        if ( bestPlan.cost() <= GOOD_ENOUGH_PLAN_COST ) break;
      }
    }

    return new AltSelectPlan(bestState, bestPlan);
  }

  public long size(Object state) {
    if ( state == null ) return 0;
    Object[] s = (Object[]) state;
    return s.length > 0 ? delegates_.get(0).size(s[0]) : 0;
  }
}
